wakuv2/nwaku.go

```diff
@@ -2377,547 +2329,737 @@ func wakuNew(nodeKey *ecdsa.PrivateKey,
-	return nil, errors.New(errMsg)
 }
 
-func (self *Waku) WakuStart() error {
-
-	var resp = C.allocResp()
-	defer C.freeResp(resp)
-	C.cGoWakuStart(self.wakuCtx, resp)
-
-	if C.getRet(resp) == C.RET_OK {
-		return nil
-	}
-	errMsg := "error WakuStart: " + C.GoStringN(C.getMyCharPtr(resp), C.int(C.getMyCharLen(resp)))
-	return errors.New(errMsg)
+//export globalEventCallback
+func globalEventCallback(callerRet C.int, msg *C.char, len C.size_t, userData unsafe.Pointer) {
+	// This is shared among all Golang instances
+	// TODO-nwaku
+	// self := Waku{wakuCtx: userData}
+	// self.MyEventCallback(callerRet, msg, len)
 }
 
-func (self *Waku) WakuStop() error {
-	var resp = C.allocResp()
-	defer C.freeResp(resp)
-	C.cGoWakuStop(self.wakuCtx, resp)
-
-	if C.getRet(resp) == C.RET_OK {
-		return nil
-	}
-	errMsg := "error WakuStop: " + C.GoStringN(C.getMyCharPtr(resp), C.int(C.getMyCharLen(resp)))
-	return errors.New(errMsg)
+func (self *Waku) MyEventCallback(callerRet C.int, msg *C.char, len C.size_t) {
+	fmt.Println("Event received:", C.GoStringN(msg, C.int(len)))
 }
-func (self *Waku) WakuDestroy() error {
-	var resp = C.allocResp()
-	defer C.freeResp(resp)
-	C.cGoWakuDestroy(self.wakuCtx, resp)
 
-	if C.getRet(resp) == C.RET_OK {
-		return nil
-	}
-	errMsg := "error WakuDestroy: " + C.GoStringN(C.getMyCharPtr(resp), C.int(C.getMyCharLen(resp)))
-	return errors.New(errMsg)
+type response struct {
+	err   error
+	value any
 }
 
-func (self *Waku) StartDiscV5() error {
-	var resp = C.allocResp()
-	defer C.freeResp(resp)
-	C.cGoWakuStartDiscV5(self.wakuCtx, resp)
+type WakuNode struct {
+	wakuCtx unsafe.Pointer
+	cancel  context.CancelFunc
+	newCh   chan *request
+}
+
+type requestType int
+
+const (
+	requestTypeNew requestType = iota + 1
+	requestTypePing
+	requestTypeStart
+	requestTypeRelayPublish
+	requestTypeStoreQuery
+	requestTypePeerID
+	requestTypeStop
+	requestTypeDestroy
+	requestTypeStartDiscV5
+	requestTypeStopDiscV5
+	requestTypeVersion
+	requestTypeRelaySubscribe
+	requestTypeRelayUnsubscribe
+	requestTypePeerExchangeRequest
+	requestTypeConnect
+	requestTypeDialPeerByID
+	requestTypeListenAddresses
+	requestTypeENR
+	requestTypeListPeersInMesh
+	requestTypeGetConnectedPeers
+	requestTypeGetPeerIDsFromPeerStore
+	requestTypeGetPeerIDsByProtocol
+	requestTypeDisconnectPeerByID
+	requestTypeDnsDiscovery
+	requestTypeDialPeer
+	requestTypeGetNumConnectedRelayPeers
+)
 
-	if C.getRet(resp) == C.RET_OK {
-		return nil
-	}
-	errMsg := "error WakuStartDiscV5: " + C.GoStringN(C.getMyCharPtr(resp), C.int(C.getMyCharLen(resp)))
-	return errors.New(errMsg)
+type request struct {
+	id         string
+	reqType    requestType
+	input      any
+	responseCh chan response
 }
 
-func (self *Waku) StopDiscV5() error {
-	var resp = C.allocResp()
-	defer C.freeResp(resp)
-	C.cGoWakuStopDiscV5(self.wakuCtx, resp)
+func NewWakuNode(ctx context.Context, config *WakuConfig) (*WakuNode, error) {
```

Comment:
It seems not needed to be public
```suggestion
func newWakuNode(ctx context.Context, config *WakuConfig) (*WakuNode, error) {
```